File: src/native/network.ts
import ffi from "./ffi.ts";
import { Layer } from "./layer.ts";
import { Matrix } from "./matrix.ts";

const {
  network_free,
  network_new,
  network_predict,
  network_init,
} = ffi;

const NetworkFinalizer = new FinalizationRegistry(
  (network: Deno.PointerValue) => {
    // network_free(network);
  },
);

enum C_COST {
  crossentropy = 0,
  mse = 1,
}

export type Cost = keyof typeof C_COST;

export interface NetworkConfig {
  inputSize: number;
  layers: Layer[];
  cost: Cost;
}

export class Network {
  #ptr: Deno.PointerValue;
  #token: { ptr: Deno.PointerValue } = { ptr: 0 };

  get unsafePointer() {
    return this.#ptr;
  }

  constructor(config: NetworkConfig) {
    this.#ptr = network_new(
      config.inputSize,
      new BigUint64Array(config.layers.map((e) => BigInt(e.unsafePointer))),
      config.layers.length,
      C_COST[config.cost],
    );
    this.#token.ptr = this.#ptr;
    NetworkFinalizer.register(this, this.#ptr, this.#token);
  }

  init(inputSize: number, batchSize: number): void {
    network_init(this.#ptr, inputSize, batchSize);
  }

  predict(input: Matrix<"f32">): Matrix<"f32"> {
    return new Matrix(network_predict(this.#ptr, input.unsafePointer));
  }

  free(): void {
    if (this.#ptr) {
      network_free(this.#ptr);
      this.#ptr = 0;
      NetworkFinalizer.unregister(this.#token);
    }
  }
}

File: src/native/network.c
#include "network.h"
#include "stdio.h"
#include "stdlib.h"

Network* network_new(unsigned int inputSize, Layer** layers, unsigned int numLayers, CostFunction cost) {
  Network* network = malloc(sizeof(Network));
  network->inputSize = inputSize;
  network->numLayers = numLayers;
  network->layers = malloc(sizeof(Layer*) * numLayers);
  for (int i = 0; i < numLayers; i++) {
    network->layers[i] = layers[i];
  }
  network->cost = cost_from_type(cost);
  return network;
}

void network_free(Network* network) {
  // Not owned by Network
  // for (unsigned int i = 0; i < network->numLayers; i++) {
  //   layer_free(network->layers[i]);
  // }
  free(network->layers);
  free(network);
}

void network_init(Network* network, unsigned int inputSize, unsigned int batches) {
  for (unsigned int i = 0; i < network->numLayers; i++) {
    Layer* layer = network->layers[i];
    layer_init(layer, inputSize, batches);
    inputSize = layer->outputSize;
  }
}

Matrix* network_feed_forward(Network* network, Matrix* input) {
  for (unsigned int i = 0; i < network->numLayers; i++) {
    input = layer_feed_forward(network->layers[i], input);
  }
  return input;
}

Matrix* network_predict(Network* network, float* xs, unsigned int xslen) {
  for (int i = 0; i < network->numLayers; i++) {
    layer_reset(network->layers[i], 1);
  }
  Matrix* input = matrix_new_from_array_zero_copy(1, xslen, TYPE_F32, xs);
  Matrix* result = network_feed_forward(network, input);
  input->data = NULL;
  matrix_free(input);
  return result;
}
